### PR TITLE
Avoid NPE in DiskThresholdMonitor

### PR DESCRIPTION
**Pull request by pxsalehi**
- Process new `ClusterInfo` only when the cluster state is recovered, since we need to know the indices/nodes in the cluster for this. 
- Avoid missing node IDs that could result from a `REPLACE` shutdown metadata. 

Both of these could lead to NPEs which is not fatal in 8.7/8.8, but could lead to unresolved listeners on 7.17.

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
Fixes look good but I think we should have some tests too - see `DiskThresholdMonitorTests`

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitorTests.java [hunk 1]
```diff
@@ -1281,6 +1283,90 @@ public void testDiskMonitorLoggingWithMaxHeadrooms() throws IllegalAccessExcepti
...
+        Map<String, DiskUsage> diskUsages = new HashMap<>();
+        diskUsages.put("node1", new DiskUsage("node1", "node1", "/foo/bar", 100, between(0, 4)));
+        diskUsages.put("node2", new DiskUsage("node2", "node2", "/foo/bar", 100, between(0, 4)));
+        final ClusterInfo clusterInfo = clusterInfo(diskUsages);
+        var result = runDiskThresholdMonitor(clusterState, clusterInfo);
+        assertTrue(result.v1()); // reroute on new nodes
+        assertEquals(Set.of("test"), result.v2());
+
+        final ClusterState blockedClusterState = ClusterState.builder(clusterState)
+            .blocks(ClusterBlocks.builder().addGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK).build())
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitorTests.java:1335]: nit: we should also have an empty routing table in this state (the routing table is initialised when removing `GatewayService.STATE_NOT_RECOVERED_BLOCK`)

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitorTests.java [hunk 2]
```diff
@@ -1281,6 +1283,90 @@ public void testDiskMonitorLoggingWithMaxHeadrooms() throws IllegalAccessExcepti
...
+            );
+        }
+        Metadata metadata = metadataBuilder.build();
+        RoutingTable routingTable = RoutingTable.builder(TestShardRoutingRoleStrategies.DEFAULT_ROLE_ONLY)
+            .addAsNew(metadata.index("test"))
+            .build();
+        DiscoveryNodes.Builder discoveryNodes = DiscoveryNodes.builder()
+            .add(newNormalNode("node1", "node1"))
+            .add(newNormalNode("node2", "node2"));
+        // node3 which is to replace node1 may or may not bee in the cluster
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitorTests.java:1314]: typo :)

```suggestion
        // node3 which is to replace node1 may or may not be in the cluster
```

**Review by DaveCTurner — APPROVED** [review]
LGTM, just a couple of very small nits

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitorTests.java (hunk 2 above)
Comment (pxsalehi) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitorTests.java:1314]: 🐝

**Pull request merged**